### PR TITLE
spec(consensus): update description of consensus parameters for PBTS

### PR DESCRIPTION
**Pull request by cason**
Addresses the Specification section of https://github.com/cometbft/cometbft/issues/2063.

Updates the consensus specs of [time](https://github.com/cometbft/cometbft/blob/cason/2063-spec/spec/consensus/time.md) and [BFT Time](https://github.com/cometbft/cometbft/blob/cason/2063-spec/spec/consensus/bft-time.md), as well as the spec of [data structures](https://github.com/cometbft/cometbft/blob/cason/2063-spec/spec/core/data_structures.md#consensusparams).

This PR also aligns the description of consensus parameters from the specs with the comments in the Go and the Protocol Buffers definitions of them. Protocol buffers are regenerated because of a change in the order of fields. 

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: proto/cometbft/types/v1/params.proto [hunk 1]
```diff
@@ -81,39 +88,45 @@ message HashedParams {
...
+// 
+// These parameters are part of the Proposer-Based Timestamps (PBTS) algorithm.
+// For more information on the relationship of the synchrony parameters to
+// block timestamps validity, refer to the PBTS specification:
 message SynchronyParams {
-  // message_delay bounds how long a proposal message may take to reach all validators on a network
-  // and still be considered valid.
-  google.protobuf.Duration message_delay = 1
+  // Bound for how skewed a proposer's clock may be from any validator on the
+  // network while still producing valid proposals.
```
Comment (lasarojc) [inline proto/cometbft/types/v1/params.proto:98]:
```suggestion
  // network while still producing valid proposals, in ms.
```

File: proto/cometbft/types/v1/params.proto [hunk 2]
```diff
@@ -81,39 +88,45 @@ message HashedParams {
...
+  google.protobuf.Duration precision = 1
       [(gogoproto.stdduration) = true];
-  // precision bounds how skewed a proposer's clock may be from any validator
-  // on the network while still producing valid proposals.
-  google.protobuf.Duration precision = 2
+  // Bound for how long a proposal message may take to reach all validators on
+  // a network and still be considered valid.
```
Comment (lasarojc) [inline proto/cometbft/types/v1/params.proto:102]:
```suggestion
  // a network and still be considered valid, in ms.
```

File: proto/cometbft/types/v1/params.proto [hunk 3]
```diff
@@ -81,39 +88,45 @@ message HashedParams {
...
+  // First height during which vote extensions will be enabled.
+  // 
+  // During the specified height, and for all subsequent heights, precommit
+  // messages that do not contain valid extension data will be considered
+  // invalid. Prior to this height, vote extensions will not be used or
+  // accepted by validators on the network.
   //
   // Once enabled, vote extensions will be created by the application in ExtendVote,
   // passed to the application for validation in VerifyVoteExtension and given
   // to the application to use when proposing a block during PrepareProposal.
```
Comment (lasarojc) [inline proto/cometbft/types/v1/params.proto:118]: , in the following height.

File: proto/cometbft/types/v1/params.proto [hunk 4]
```diff
@@ -81,39 +88,45 @@ message HashedParams {
...
-  // vote_extensions_enable_height configures the first height during which
-  // vote extensions will be enabled. During this specified height, and for all
-  // subsequent heights, precommit messages that do not contain valid extension data
-  // will be considered invalid. Prior to this height, vote extensions will not
-  // be used or accepted by validators on the network.
+  // First height during which vote extensions will be enabled.
+  // 
+  // During the specified height, and for all subsequent heights, precommit
+  // messages that do not contain valid extension data will be considered
+  // invalid. Prior to this height, vote extensions will not be used or
```
Comment (lasarojc) [inline proto/cometbft/types/v1/params.proto:113]:
```suggestion
  // invalid. Prior to this height, or when this height is set to 0, vote extensions will not be used or
```

File: spec/consensus/bft-time.md [hunk 5]
```diff
@@ -47,9 +47,10 @@ Consider the following example:
   The total voting power is 70, so we assume that the faulty validators have at
   most 23 of voting power (since `N < 3F`, where `N` is the total voting
   power and `F` is the maximum voting power of faulty validators).
-- We have the following `Precommit` messages in some `LastCommit` field (we
-  ignore all fields except the `Timestamp` field): (p1, 100), (p2, 98), (p3, 1000), (p4, 500).
-  We assume that p3 and p4 are faulty validators.
+- We have the following `Precommit` messages in some `Commit` field (notice that we
+  ignore all fields except the validator name and the `Timestamp` field): (p1, 100), (p2, 98), (p3, 1000), (p4, 500).
+  We assume that p3 and p4 are faulty validators, as they propose block times
+  far much higher (fart in the future) than p1 and p2.
```
Comment (lasarojc) [inline spec/consensus/bft-time.md:53]:
```suggestion
  far much higher (far in the future) than p1 and p2.
```

File: spec/consensus/time.md [hunk 6]
```diff
@@ -16,15 +16,19 @@ The Time produced by CometBFT satisfies the following properties:
...
+  to arbitrarily increase or decrease the block Time.
+  In other words, the Time of blocks should be defined by correct nodes.
 
 In addition, the Time produced by CometBFT is expected, by external observers, to provide:
 
 - **Relation to real time**: block times bear some resemblance to real time.
   In other words, block times should represent, within some reasonable accuracy,
-  the actual time at which blocks were produced.
+  the actual clock time at which blocks were produced.
+  More formally, lets `t` be the clock time at which a block with header `H`
```
Comment (lasarojc) [inline spec/consensus/time.md:28]: The use of real time in these specs should be clarified to mean the external source of time to which these nodes synchronize. It is not really real time, but a time reference, since there is no real "real time".

Another way would be to consider the real time the median validator clock. This way it may be used to express better the guarantees of BFT Time as well.

```suggestion
  More formally, lets `t` be the clock time at which a block with header `H`
```

File: spec/consensus/time.md [hunk 7]
```diff
@@ -33,20 +37,53 @@ CometBFT implements two algorithms for computing block times:
...
+inaccurate block times because they rely on inaccurate local clocks.
+
+For more details, refer to the specification of [BFT Time][bft-time] and [Proposer-Based Timestamps][pbts-spec].
+
+## Adopting PBTS
+
+The Proposer-Based Timestamp (PBTS) algorithm is the recommended algorithm for
+producing block times.
+
+As of CometBFT `v1.x`, however, PBTS is not enabled by default, either for new
```
Comment (lasarojc) [inline spec/consensus/time.md:66]:
```suggestion
As of CometBFT `v1.x`, however, PBTS is not enabled by default, neither for new
```

File: spec/core/data_structures.md [hunk 8]
```diff
@@ -124,22 +124,22 @@ The steps to validate a new block are:
...
+|-------------------|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
```
Comment (lasarojc) [inline spec/core/data_structures.md:128]:
```suggestion
|-----|------|-----|-----|
```

File: spec/core/data_structures.md [hunk 9]
```diff
@@ -466,62 +466,101 @@ func SumTruncated(bz []byte) []byte {
...
-|--------------|-------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------|
-| max_bytes    | int64 | Max size of a block, in bytes.                                                                                                                                                                              | 1            |
-| max_gas      | int64 | Max sum of `GasWanted` in a proposed block. NOTE: blocks that violate this may be committed if there are Byzantine proposers. It's the application's responsibility to handle this when processing a block! | 2            |
+| Name      | Type  | Description                                             | Field Number |
+|-----------|-------|---------------------------------------------------------|:------------:|
+| max_bytes | int64 | Maximum size of a block, in bytes.                      | 1            |
+| max_gas   | int64 | Maximum gas wanted by transactions included in a block. | 2            |
+
+The `max_bytes` parameter must be greater or equal to -1.
```
Comment (lasarojc) [inline spec/core/data_structures.md:493]:
```suggestion
The `max_bytes` parameter must be greater or equal to -1 and smaller or equal to 100MB.
```

File: spec/core/data_structures.md [hunk 10]
```diff
@@ -466,62 +466,101 @@ func SumTruncated(bz []byte) []byte {
...
+### FeatureParams
+
+| Name                          | Type  | Description                                                       | Field Number |
+|-------------------------------|-------|-------------------------------------------------------------------|:------------:|
+| vote_extensions_enable_height | int64 | First height during which vote extensions will be enabled.        | 1            |
+| pbts_enable_height            | int64 | Height at which Proposer-Based Timestamps (PBTS) will be enabled. | 2            |
+
+From the configured height, and for all subsequent heights, the corresponding
+feature will be enabled.
+Cannot be set to heights lower or equal to the current blockchain height.
```
Comment (lasarojc) [inline spec/core/data_structures.md:551]:
```suggestion
Cannot be set to heights lower or equal to the current blockchain height.
A value of -1 indicates that the feature is disabled.
```

File: spec/core/data_structures.md [hunk 11]
```diff
@@ -466,62 +466,101 @@ func SumTruncated(bz []byte) []byte {
...
-| message_delay | [google.protobuf.Duration](https://protobuf.dev/reference/protobuf/google.protobuf/#duration) | Bound for how long a proposal message may take to reach all validators on a network and still be considered valid.      | 2            |
+| Name          | Type                                       | Description                                                                                                             | Field Number |
+|---------------|--------------------------------------------|-------------------------------------------------------------------------------------------------------------------------|:------------:|
+| precision     | [google.protobuf.Duration][proto-duration] | Bound for how skewed a proposer's clock may be from any validator on the network while still producing valid proposals. | 1            |
+| message_delay | [google.protobuf.Duration][proto-duration] | Bound for how long a proposal message may take to reach all validators on a network and still be considered valid.      | 2            |
```
Comment (lasarojc) [inline spec/core/data_structures.md:558]:
```suggestion
| precision     | [google.protobuf.Duration][proto-duration] | Bound for how skewed a proposer's clock may be from any validator on the network while still producing valid proposals, in ms. | 1            |
| message_delay | [google.protobuf.Duration][proto-duration] | Bound for how long a proposal message may take to reach all validators on a network and still be considered valid, in ms.      | 2            |
```

**Review by lasarojc — APPROVED** [review]
Just have a quibble  comment on the definition of real time that might have repercussions.

File: proto/cometbft/types/v1/params.proto (hunk 1 above)
Comment (cason) [reply proto/cometbft/types/v1/params.proto:98]: It is not. It is a `time.Duration`, an `int64`, given in nanos: https://pkg.go.dev/time#Duration.

If it said otherwise in other place, please let me known.

File: proto/cometbft/types/v1/params.proto (hunk 2 above)
Comment (cason) [reply proto/cometbft/types/v1/params.proto:102]: It is not. It is a `time.Duration`, an `int64`, given in nanos: https://pkg.go.dev/time#Duration.

If it said otherwise in other place, please let me known.

File: proto/cometbft/types/v1/params.proto [hunk 12]
```diff
@@ -21,55 +21,62 @@ message ConsensusParams {
...
 }
 
-// BlockParams contains limits on the block size.
+// BlockParams define limits on the block size and gas.
 message BlockParams {
-  // Max block size, in bytes.
-  // Note: must be greater than 0
+  // Maximum size of a block, in bytes.
+  //
+  // Must be greater or equal to -1. If set to -1, the limit is the maximum
```
[the remark this answers is not in the document]
Comment (cason) [reply proto/cometbft/types/v1/params.proto:28]: Addressed.

File: spec/consensus/time.md (hunk 6 above)
Comment (cason) [reply spec/consensus/time.md:28]: Yes, this is odd.

File: spec/core/data_structures.md (hunk 10 above)
Comment (cason) [reply spec/core/data_structures.md:551]: -1 or 0?

File: spec/core/data_structures.md (hunk 11 above)
Comment (cason) [reply spec/core/data_structures.md:558]: Not in ms.

File: spec/core/data_structures.md (hunk 8 above)
Comment (cason) [reply spec/core/data_structures.md:128]: Not sure if I understood.

File: proto/cometbft/types/v1/params.proto (hunk 2 above)
Comment (cason) [reply proto/cometbft/types/v1/params.proto:105]: Fixed this misunderstanding in #2376

File: spec/core/data_structures.md (hunk 8 above)
Comment (lasarojc) [reply spec/core/data_structures.md:128]: Just a suggestion.
These dashes just clutter the file, but have not effect on the rendered table.

File: spec/core/data_structures.md (hunk 10 above)
Comment (lasarojc) [reply spec/core/data_structures.md:552]: 0.
My mind was already in the test code.
